File: internal/naginata/engine.go
package naginata

import "time"

// Emission はエンジンが確定した1文字（または拗音などの1単位）を表す。
type Emission struct {
	Text  string
	Chord KeySet
}

// Engine はキー押下の列から薙刀式のかなを確定する。
//
// 端末入力ではキーリリースを検出できないため、QMKの「リリースで確定」の
// 代わりにタイミングウィンドウを使う。ウィンドウ内に押されたキーを
// 同時押しとみなし、QMKと同じ最長一致の貪欲法で確定する。
type Engine struct {
	window   time.Duration
	buf      []Key
	deadline time.Time
	armed    bool
	presses  int
}

// NewEngine はウィンドウ幅を指定してエンジンを作る。
func NewEngine(window time.Duration) *Engine {
	return &Engine{window: window}
}

// Presses は起動からの総打鍵数を返す。
func (e *Engine) Presses() int { return e.presses }

// Pending は未確定のキーがあるかを返す。
func (e *Engine) Pending() bool { return len(e.buf) > 0 }

// Press はキー押下を処理し、確定したかなを返す。
func (e *Engine) Press(k Key, now time.Time) []Emission {
	e.presses++

	var out []Emission
	// 同じキーの連打は同時押しになり得ないので、先にバッファを確定する。
	if e.comb(len(e.buf)).Has(k) {
		out = append(out, e.flushAll()...)
	}
	e.buf = append(e.buf, k)

	for len(e.buf) > 0 {
		nc, complete := e.candidates()
		if nc == 0 {
			// もう組み合わせが伸びないので、先頭から確定できる分を確定する。
			out = append(out, e.typeOnce()...)
			continue
		}
		if nc == 1 && complete {
			out = append(out, e.flushAll()...)
		}
		break
	}

	if len(e.buf) > 0 {
		e.deadline = now.Add(e.window)
		e.armed = true
	} else {
		e.armed = false
	}
	return out
}

// Flush はウィンドウ経過後に未確定のキーを確定する。定期的に呼ぶ。
func (e *Engine) Flush(now time.Time) []Emission {
	if !e.armed || now.Before(e.deadline) {
		return nil
	}
	e.armed = false
	return e.flushAll()
}

// comb はバッファ先頭 n キーの組み合わせを返す。
func (e *Engine) comb(n int) KeySet {
	var s KeySet
	for _, k := range e.buf[:n] {
		s |= 1 << k
	}
	return s
}

// candidates はバッファ全体を含む組み合わせの候補数を返す。
// complete は候補が一つでかつ全キーが押されていることを示す。
func (e *Engine) candidates() (count int, complete bool) {
	comb := e.comb(len(e.buf))
	var hit KeySet
	for _, entry := range Table {
		if entry.Keys&comb == comb {
			count++
			hit = entry.Keys
		}
	}
	complete = count == 1 && len(e.buf) >= hit.Count()
	return count, complete
}

// typeOnce はバッファ先頭からの最長一致で1単位を確定する。
// どの長さでも一致しなければ先頭の1キーを捨てる。
func (e *Engine) typeOnce() []Emission {
	for nt := len(e.buf); nt > 0; nt-- {
		comb := e.comb(nt)
		for _, entry := range Table {
			if entry.Keys == comb {
				e.buf = e.buf[nt:]
				return []Emission{{Text: entry.Text, Chord: comb}}
			}
		}
	}
	e.buf = e.buf[1:]
	return nil
}

// flushAll はバッファが空になるまで確定を繰り返す。
func (e *Engine) flushAll() []Emission {
	var out []Emission
	for len(e.buf) > 0 {
		out = append(out, e.typeOnce()...)
	}
	return out
}

File: web/src/lib/engine.ts
// 薙刀式の同時押し判定エンジン。
//
// Go版 (ターミナル) はキーリリースを検出できずタイミングウィンドウで
// 近似していたが、ブラウザでは keydown/keyup が取れるので、QMK実装
// (eswai/qmk_firmware naginata_v15) の「リリースで確定」を忠実に移植する。
// 遅延がなく、連続シフト・連続濁音もそのまま動く。

import { count, has, type Key, type KeySet } from "./keys";
import { TABLE } from "./table";

export interface Emission {
  text: string;
}

// 連続シフトのとき押しっぱなしを引き継ぐキー (QMK の shifted lookup 相当)。
const F = 13, J = 16, V = 23, M = 26;
const SPACE = 30;
const STICKY: Key[] = [SPACE, F, J, V, M];

export class Engine {
  private buf: Key[] = []; // 押された順の未確定キー
  private held: KeySet = 0; // いま物理的に押されているキー
  presses = 0;

  press(key: Key): Emission[] {
    this.presses++;
    this.held |= 1 << key;

    // スペースも普通のキーとして組み合わせに参加させる (後入れシフト許容)。
    // Go版と同じ判断で、かなを打った直後のシフトも同時押しとして拾う。
    this.buf.push(key);
    const { n, complete } = this.candidates();
    if (n === 0 || (n === 1 && complete)) {
      return this.typeOnce();
    }
    return [];
  }

  release(key: Key): Emission[] {
    this.held &= ~(1 << key);
    if (this.buf.length > 0) {
      return this.typeOnce();
    }
    return [];
  }

  // 未確定のキーを捨てる。単語の切り替え時に呼ぶ。
  reset(): void {
    this.buf = [];
  }

  // 押しっぱなしの記録も消す。フォーカスが外れて keyup を
  // 取り逃がしたときに呼ぶ。
  clearHeld(): void {
    this.held = 0;
    this.buf = [];
  }

  private combOf(n: number): KeySet {
    let s = 0;
    for (let i = 0; i < n; i++) s |= 1 << this.buf[i];
    return s;
  }

  // バッファ全体を含む候補の数。complete は候補が一つでかつ
  // 全キーが押されていることを示す。
  private candidates(): { n: number; complete: boolean } {
    const comb = this.combOf(this.buf.length);
    let n = 0;
    let hit = 0;
    for (const e of TABLE) {
      if ((e.keys & comb) === comb) {
        n++;
        hit = e.keys;
      }
    }
    return { n, complete: n === 1 && this.buf.length >= count(hit) };
  }

  // バッファ先頭からの最長一致で1単位を確定する。
  // 連続シフト: 押しっぱなしのシフト・濁音キーを組み合わせに足して先に探す。
  // どの長さでも一致しなければ先頭の1キーを捨てる。
  private typeOnce(): Emission[] {
    for (let nt = this.buf.length; nt > 0; nt--) {
      const comb = this.combOf(nt);

      let sticky = comb;
      for (const k of STICKY) {
        if (has(this.held, k)) sticky |= 1 << k;
      }
      if (sticky !== comb) {
        const e = TABLE.find((e) => e.keys === sticky);
        if (e) {
          this.buf = this.buf.slice(nt);
          return [{ text: e.text }];
        }
      }
      const e = TABLE.find((e) => e.keys === comb);
      if (e) {
        this.buf = this.buf.slice(nt);
        return [{ text: e.text }];
      }
    }
    this.buf.shift();
    return [];
  }
}
